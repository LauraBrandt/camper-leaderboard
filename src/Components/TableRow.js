import React from 'react';
import PropTypes from 'prop-types';

const TableRow = ({ user, index }) => {
  return(
    <tr className="row">
      <td>{index+1}</td>
      <td>
        <a href={`https://www.freecodecamp.org/${user.username}`} className="user-info">
          <img src={user.img} alt={`${user.username} avatar`}/>
          <p>{user.username}</p>
        </a>
      </td>
      <td>{user.recent}</td>
      <td>{user.alltime}</td>
    </tr>
  );
}

TableRow.propTypes = {
  index: PropTypes.number.isRequired,
  user: PropTypes.shape({
    username: PropTypes.string.isRequired,
    recent: PropTypes.number.isRequired,
    alltime: PropTypes.number.isRequired,
    img: PropTypes.string
  })
}

export default TableRow;
