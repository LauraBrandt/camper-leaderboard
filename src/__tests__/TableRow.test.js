import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import TableRow from '../Components/TableRow';

describe('TableRow', () => {
  let mockProps = {
    user: {
      username: 'camper',
      recent: 5,
      alltime: 10,
      img: "#"
    },
    index: 0
  };
  
  it('renders and matches snapshot', () => {
    const renderedComponent = renderer.create(<TableRow {...mockProps}/>);
    const tree = renderedComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  describe('content', () => {
    let component;
    const resetProps = () => mockProps = {
      user: {
        username: 'camper',
        recent: 5,
        alltime: 10,
        img: "#"
      },
      index: 0
    }

    beforeEach(() => {
      resetProps();
      component = shallow(<TableRow {...mockProps}/>);
    });

    it('has 4 columns', () => {
      expect(component.find('tr td')).toHaveLength(4);
    });

    it('the first column contains the prop index + 1', () => {
      expect(component.find('td').at(0).text()).toBe((mockProps.index + 1).toString());
    });

    it('the second column contains a link with href containing the username from prop user', () => {
      const column = component.find('td').at(1);
      const link = column.find('a');
      expect(link).toHaveLength(1);
      expect(link.prop('href')).toContain(mockProps.user.username);
    });
    
    it('the previous link contains an image with src and alt that contains the username from prop user', () => {
      const container = component.find('td').at(1).find('a');
      
      const image = container.find('img');
      expect(image).toHaveLength(1);
      expect(image.prop('src')).toBe(mockProps.user.img);
      expect(image.prop('alt')).toContain(mockProps.user.username);
    });

    it('the previous link contains the username text from prop user', () => {
      const container = component.find('td').at(1).find('a');
      expect(container.text()).toContain(mockProps.user.username);
    });
    
    it('the third column contains the recent points from prop user', () => {
      expect(component.find('td').at(2).text()).toBe((mockProps.user.recent).toString());
    });
    
    it('the fourth column contains the all time points from prop user', () => {
      expect(component.find('td').at(3).text()).toBe((mockProps.user.alltime).toString());
    });
  });
});