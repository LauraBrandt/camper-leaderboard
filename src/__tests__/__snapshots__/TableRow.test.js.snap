// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow renders and matches snapshot 1`] = `
<tr
  className="row"
>
  <td>
    1
  </td>
  <td>
    <a
      className="user-info"
      href="https://www.freecodecamp.org/camper"
    >
      <img
        alt="camper avatar"
        src="#"
      />
      <p>
        camper
      </p>
    </a>
  </td>
  <td>
    5
  </td>
  <td>
    10
  </td>
</tr>
`;
